subprojects/instant-execution/src/main/kotlin/org/gradle/instantexecution/serialization/codecs/Codecs.kt

```diff
@@ -0,0 +1,146 @@
+import org.gradle.api.invocation.Gradle
+import org.gradle.api.model.ObjectFactory
+import org.gradle.api.tasks.util.internal.PatternSpecFactory
+import org.gradle.instantexecution.serialization.DecodingProvider
+import org.gradle.instantexecution.serialization.Encoding
+import org.gradle.instantexecution.serialization.EncodingProvider
+import org.gradle.instantexecution.serialization.IsolateContext
+import org.gradle.instantexecution.serialization.ReadContext
+import org.gradle.instantexecution.serialization.WriteContext
+import org.gradle.instantexecution.serialization.bindings
+import org.gradle.instantexecution.serialization.logUnsupported
+import org.gradle.instantexecution.serialization.ownerProjectService
+import org.gradle.instantexecution.serialization.singleton
+import org.gradle.internal.reflect.Instantiator
+import org.gradle.internal.serialize.BaseSerializerFactory.BOOLEAN_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.BYTE_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.DOUBLE_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.FILE_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.FLOAT_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.INTEGER_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.LONG_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.SHORT_SERIALIZER
+import org.gradle.internal.serialize.BaseSerializerFactory.STRING_SERIALIZER
+import org.gradle.internal.serialize.SetSerializer
+import kotlin.reflect.KClass
+
+
+class Codecs(
+    directoryFileTreeFactory: DirectoryFileTreeFactory,
+    private val fileCollectionFactory: FileCollectionFactory,
+    private val fileResolver: FileResolver,
+    private val instantiator: Instantiator,
+    private val objectFactory: ObjectFactory,
+    private val patternSpecFactory: PatternSpecFactory,
+    private val filePropertyFactory: FilePropertyFactory
+) : EncodingProvider, DecodingProvider {
+
+    private
+    val fileSetSerializer = SetSerializer(FILE_SERIALIZER)
+
+    private
+    val bindings = bindings {
+
+        bind(STRING_SERIALIZER)
+        bind(BOOLEAN_SERIALIZER)
+        bind(INTEGER_SERIALIZER)
+        bind(SHORT_SERIALIZER)
+        bind(LONG_SERIALIZER)
+        bind(BYTE_SERIALIZER)
+        bind(FLOAT_SERIALIZER)
+        bind(DOUBLE_SERIALIZER)
+        bind(FileTreeCodec(fileSetSerializer, directoryFileTreeFactory))
+        bind(FILE_SERIALIZER)
+        bind(ClassCodec)
+
+        bind(listCodec)
+        bind(setCodec)
+
+        // Only serialize certain Map implementations for now, as some custom types extend Map (eg DefaultManifest)
+        bind(HashMap::class, mapCodec)
+
+        bind(LoggerCodec)
+
+        bind(FileCollectionCodec(fileSetSerializer, fileCollectionFactory))
+        bind(ArtifactCollectionCodec)
+
+        bind(singleton(objectFactory))
```

Comment:
We should really use the owner project for these services too. It's not entirely correct either way (always using global scoped or always using project scoped), but the project services are more likely to have been used in the original build than global services. We can make this more accurate later, if necessary, or plug the global service leakages in the original build.